### PR TITLE
add ide-specific settings to .gitignore and simply implement 7z

### LINKED ISSUE #164: Support `.7z`
The title sums up the first part of the request pretty well, 7z is a pretty popular archive format, but 7z supports several different compression formats it supports, so I think it would also be good to add an option like `--format`/`-f` to specify a compression format independently of the file name, which would be useful in general too

### PR DESCRIPTION
**Pull request by MisileLab**
fix #164

## Description
implement 7z and add .vscode and .idea to .gitignore

### todo
- [x] implement 7z list
- [x] return error not using unwrap

### CODE REVIEW

**Comment by marcospb19** [comment]
Thanks for your contribution! A review will come once one of us finds the time to take a look.

**Review by marcospb19 — CHANGES_REQUESTED** [review]
Thanks for your PR, there are some changes needed for it to be merged.

In the title you wrote "simply implement 7z", so maybe your plan was to leave this as it is right now, a simpler implementation, if that's the case, I'll ask you to leave the PR open so that anyone can pick from where you left it and finish it, otherwise, you're more than welcome to finish it yourself.

File: src/commands/list.rs [hunk 1]
```diff
@@ -78,6 +79,15 @@ pub fn list_archive_contents(
 
             Box::new(crate::archive::zip::list_archive(zip_archive))
         }
+        SevenZip => {
+            let a = RefCell::new(Vec::new());
```
Comment (marcospb19) [inline src/commands/list.rs:83]: If https://github.com/dyz1990/sevenz-rust/pull/17 gets merged, you can remove the `RefCell` wrap.

File: src/commands/mod.rs [hunk 2]
```diff
@@ -21,6 +21,21 @@ use crate::{
...
+        let entry = entry?;
+        let filetype = entry.file_type()?;
+        if filetype.is_dir() {
+            copy_recursively(entry.path(), destination.as_ref().join(entry.file_name()))?;
+        } else {
+            fs::copy(entry.path(), destination.as_ref().join(entry.file_name()))?;
+        }
+    }
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/mod.rs:37]: smart_unpack is rename and copy_recursively is copy, so probably need copy_recursively function.

File: src/commands/decompress.rs [hunk 3]
```diff
@@ -146,6 +146,10 @@ pub fn decompress_file(
                 return Ok(());
             }
         }
+        SevenZip => {
+            sevenz_rust::decompress_file(input_file_path, output_dir).expect("can't decompress");
+            1
```
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/decompress.rs]: fixed

File: src/commands/list.rs [hunk 4]
```diff
@@ -78,6 +79,15 @@ pub fn list_archive_contents(
...
+
+            sevenz_rust::decompress_file_with_extract_fn(archive_path, ".", |entry, _, _| {
+                a.borrow_mut().push(Ok(FileInArchive{path: entry.name().into(), is_dir: entry.is_directory()}));
+                Ok(true)
+            }).expect("failed to get 7z file list");
```
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/list.rs:88]: fixed

File: src/commands/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/mod.rs]: fixed

File: src/commands/list.rs (hunk 1 above)
Comment (MisileLab) [reply src/commands/list.rs]: fixed

File: src/commands/decompress.rs [hunk 5]
```diff
@@ -146,6 +146,16 @@ pub fn decompress_file(
...
         }
+        SevenZip => {
+            let mut count = 0;
+            sevenz_rust::decompress_file_with_extract_fn(input_file_path, output_dir, 
+            |entry, reader, dest| {
+                count += 1;
+                sevenz_rust::default_entry_extract_fn(entry, reader, dest)
+            }
+            ).map_err(|x| crate::Error::SevenzipError(x))?;
+            fs::read_dir(output_dir)?.count()
```
Comment (marcospb19) [inline src/commands/decompress.rs:157]: Maybe here you meant to put `count`

**Comment by marcospb19** [comment]
I've asked for a review from Figsoda because unfortunately I've been busy (and kinda exhausted), but he's also kinda busy so maybe this will have to wait.

File: src/commands/decompress.rs (hunk 5 above)
Comment (MisileLab) [reply src/commands/decompress.rs:157]: thank you, that's my mistake

**Review by Mic92 — APPROVED** [review]
I looked at the code and it looks pretty simple and straight forward. If I was the project maintainer I would probably also want to see some lines added in ./tests/integration.rs

**Review by marcospb19 — CHANGES_REQUESTED** [review]
This is the final change request.

As @Mic92 mentioned, it's necessary to add 7z to the integration tests file.

File: src/commands/compress.rs [hunk 6]
```diff
@@ -127,6 +128,30 @@ pub fn compress_files(
...
+                                .to_str()
+                                .unwrap()
+                                .to_string(),
+                        ),
+                        None,
+                    )
+                    .map_err(|e| crate::Error::SevenzipError(e))?;
+            }
+
+            writer.finish()?;
```
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/compress.rs]: fixed

File: src/commands/list.rs [hunk 7]
```diff
@@ -78,6 +78,19 @@ pub fn list_archive_contents(
...
+            sevenz_rust::decompress_file_with_extract_fn(archive_path, ".", |entry, _, _| {
+                a.push(Ok(FileInArchive {
+                    path: entry.name().into(),
+                    is_dir: entry.is_directory(),
+                }));
+                Ok(true)
+            })
+            .map_err(|e| crate::Error::SevenzipError(e))?;
+            Box::new(a.into_iter())
+        }
```
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/list.rs]: fixed

File: src/commands/decompress.rs [hunk 8]
```diff
@@ -146,6 +146,16 @@ pub fn decompress_file(
...
         }
+        SevenZip => {
+            let mut count = 0;
+            sevenz_rust::decompress_file_with_extract_fn(input_file_path, output_dir, 
+            |entry, reader, dest| {
+                count += 1;
+                sevenz_rust::default_entry_extract_fn(entry, reader, dest)
+            }
+            ).map_err(|x| crate::Error::SevenzipError(x))?;
+            count
```
[the remark this answers is not in the document]
Comment (MisileLab) [reply src/commands/decompress.rs]: fixed

**Comment by marcospb19** [comment]
Thanks for sticking to this! It's looking great! 

Reference for someone who wants to add integration tests: this is how Figsoda did it when he added support for Snappy:

https://github.com/ouch-org/ouch/pull/215/files#diff-eb771f54a1fdc1c330a82981d7d27d1202a5cbe3a59581ff67af5e3361f5f90e

**Comment by figsoda** [comment]
added 7z to tests, but I kept running into this error
```
thread \'main\' panicked at \'StripPrefix Failed: StripPrefixError(())\', src/archive/sevenz.rs:17:26
```

File: src/archive/sevenz.rs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                        .unwrap()
+                        .to_string(),
+                ),
+                None,
+            )
+            .map_err(crate::Error::SevenzipError)?;
+    }
+
+    writer.finish()?;
+    Ok(true)
```
Comment (marcospb19) [inline src/archive/sevenz.rs:29]: It's in line 17 of this big oneliner.

**Comment by MisileLab** [comment]
i don't know how to fix it and I have a no time for now, anyone can contribute this pull request!

**Comment by marcospb19** [comment]
@MisileLab thanks for your hard-work! I might be able to pick it up soon.

**Comment by AntoniosBarotsis** [comment]
I saw that [`tar`](https://github.com/ouch-org/ouch/blob/cd14b0038b890afc6c591c0fe666d0957149445e/src/archive/tar.rs#L43) uses [`strip_cur_dir`](https://github.com/ouch-org/ouch/blob/cd14b0038b890afc6c591c0fe666d0957149445e/src/utils/formatting.rs#L59) instead of something like `filep.strip_prefix(current_dir()?)` where the issue was in the 7z code.

Switching to that does solve the panic (I can now manually compress and decompress 7z archives from the cli) however it does not fix all the tests. Using `strip_cur_dir` in my case did **not** change the file path but instead kept it the same (again, the resulting behaviour worked as expected though). I'm not sure if this was not used on purpose or if using it is a bad idea for some reason so do let me know, if its not then I can push the commit 😅 

Another thing that was surprising was when decompressing an archive that was compressed by `ouch` I get the following log:
```sh
$ cargo r -- d tmp.7z
[INFO] Could not detect the extension of `tmp`
```

This was interesting because it did decompress correctly. The bytes [here](https://github.com/MisileLab/ouch/blob/656514acfd4f1d3911806a479d5719215e83289f/src/utils/fs.rs#L90) seem to agree with Wikipedia so not sure why it fails to detect it (but somehow still work?).




The following tests are failing:

```
test single_empty_file ... FAILED
test single_file ... FAILED
test multiple_files ... FAILED
```

**Comment by zzzsyyy** [comment]
will it work？ Whats the block so far?

**Comment by marcospb19** [comment]
> will it work？ Whats the block so far?

I believe it will. The current block is: PR is not ready yet, we're waiting for someone to pick it up and spend more time finishing it (and maybe opening another PR based on this branch, but rebased to solve conflicts).

The one who picks it up needs to debug the errors, solve them, and make the tests pass.

tip: `.tar` and `.zip` implementations can be used as a reference.

**Comment by Flat** [comment]
I've [started working on this PR](https://github.com/Flat/ouch/tree/7zip) hoping to fix the last issues and get it merged, however I seem to be stuck around the same place after re-basing. If we use the original code from this PR, 7zip will fail tests that contain any additional compression on 7z (bz, bz2 gzip, etc) this is the BadSignature error, it has the header bytes of bzip, which suggests it's being passed to the 7z decompression/unpack before it has been fully decompressed from the other algorithms.

If we try the temp file creation that rar uses, nothing is written from the reader to the temp file, which is confusing since attaching lldb I see the reader has data, but it isn't copied.

**Comment by marcospb19** [comment]
> this is the BadSignature error, it has the header bytes of bzip, which suggests it's being passed to the 7z decompression/unpack before it has been fully decompressed from the other algorithms

Does it look like a problem with our code or the `7z` crate code?

**Comment by Flat** [comment]
> @Flat
> 
> > this is the BadSignature error, it has the header bytes of bzip, which suggests it's being passed to the 7z decompression/unpack before it has been fully decompressed from the other algorithms
> 
> Does it look like a problem with our code or the `7z` crate code?

I may be barking up the wrong tree on that info actually. Doing some more testing and it looks like regular 7z compression fails, due to a symlink I have setup, the file paths passed to archive/compression algorithms are different than the env:current_dir path used in `strip_cur_dir`. Once I figure that out I'll look more into the compression/decompression chain.

**Comment by marcospb19** [comment]
Superseded by #555, thanks @MisileLab and others who gave input here.

**Pull request closed without merging**